platforms/core-configuration/file-collections/src/test/groovy/org/gradle/api/internal/file/collections/DefaultConfigurableFileCollectionSpec.groovy

```diff
@@ -2009,52 +2019,32 @@ class DefaultConfigurableFileCollectionSpec extends FileCollectionSpec {
-        when:
-        collection.from("src3")
-        collection.convention("src1", "src2")
-        then:
-        collection.from as List == ["src3"]
-    }
-
-    def "conventions can be overridden with an explicit value using #setFrom"() {
-        when:
-        collection.convention("src1", "src2")
-        collection.setFrom("src3")
-        then:
-        collection.from as List == ["src3"]
-        collection.explicit
-    }
+    def "test '#label'"(String label) {
+        given:
+        if (!(convention instanceof Wildcard)) {
+            collection.convention((Iterable) convention)
+        }
+        if (!(explicit instanceof Wildcard)) {
+            collection.setFrom((Iterable) explicit)
+        }
 
-    def "conventions can be overridden with an explicit value using #from"() {
         when:
-        collection.convention("src1", "src2")
-        collection.from("src3")
-        then:
-        collection.explicit
-        collection.from as List == ["src3"]
-    }
+        operations.each {operation -> operation.call(collection) }
 
-    def "conventions are brought back if explicit value is unset"() {
-        when:
-        collection.from("src3")
-        collection.convention("src1", "src2")
         then:
-        collection.from as List == ["src3"]
+        collection.from.flatten() as List == expected
 
-        when:
-        collection.unset()
-
-        then:
-        collection.from as List == ["src1", "src2"]
-        !collection.explicit
+        where:
+        expected        | explicit      | convention        | label                                         | operations
+        []              | _             | _                 | "no elements by default"                      | { }
+        ["src1"]        | ["src1"]      | _                 | "explicit value when set"                     | { }
+        ["src1"]        | _             | ["src1"]          | "convention used when no explicit value"      | { }
+        ["src3"]        | ["src3"]      | ["src1"]          | "explicit value overrides convention"         | { }
+        ["src1"]        | ["src3"]      | ["src1"]          | "convention used when explicit unset"         | { it.unset() }
+        ["src1", "src2"]| _             | ["src1"]          | "from() honors convention"                    | { it.from("src2") }
+        ["src2"]        | _             | _                 | "from() prevents convention"                  | { it.from("src2"); it.convention("src1") }
+        ["src2", "src3"]| ["src2"]      | ["src1"]          | "from() ignores convention"                   | { it.from("src3") }
+        ["src1", "src2"]| _             | ["src1"]          | "from() commits convention"                   | { it.from("src2"); it.unsetConvention() }
+        ["src1"]        | _             | ["src1"]          | "from() does not modify convention"           | { it.from("src2"); it.unset() }
```

Comment:
This scenario is broken, convention is modified, which is wrong

Reply:
Issue has been fixed with shallow copying of convention values when they are meant to be used as effective values.